### DISCUSSION TITLE
pretty-print a line at a time in a table?

### DISCUSSION THREAD

**Discussion by emeryberger** [Textualize/rich#965] [Q&A] [answered]
Hi all,

I'm the author of Scalene (https://github.com/emeryberger/scalene), which uses Rich (thanks for creating it!). A user pointed out that Scalene currently does pretty-printing a line at a time, which is not quite right, since it doesn't render comments correctly, and I have not yet been able to come up with a solution (issue here: https://github.com/emeryberger/scalene/issues/115).

In short, I've been trying to use `Syntax` to pretty-print the whole file at once, and then add the lines one at a time to a `Table`. When I do this, regardless of the many combinations of options I've tried so far, I find myself unable to get reasonable output. My approach was to capture the output and use the individual lines obtained via `splitlines`. The lines `print` just fine, but when they go into rows of a table, the formatting is kind of a mess, as you can see below ("line-by-line").

Pointers would be greatly appreciated - thanks in advance!

Line-by-line, broken:
<img width="1639" alt="Screen Shot 2021-01-26 at 6 26 31 PM" src="https://user-images.githubusercontent.com/1612723/105919671-0e3dd100-6004-11eb-8f0e-d1f9c992c661.png">

Current (note the incorrect rendering of the docstring at line 20):
<img width="1616" alt="Screen Shot 2021-01-26 at 6 30 28 PM" src="https://user-images.githubusercontent.com/1612723/105920084-b8b5f400-6004-11eb-8c5d-92340c1aa599.png">

**Comment by willmcgugan** [comment, marked as answer]
Hi Emery,

I think you were one of the earliest adopters of Rich.

I didn't consider this use-case so this may not be pretty. But the 'trick' is to render a single Syntax object in to lines. Rendering generates a list of [Segments](https://rich.readthedocs.io/en/latest/reference/segment.html) which can then be used to create a new renderable.

The following seems to work:

```python
from rich import print
from rich.console import Console
from rich.syntax import Syntax
from rich.table import Table

code = '''
class SyntaxTheme(ABC):
    """Base class for a syntax theme."""

    @abstractmethod
    def get_style_for_token(self, token_type: TokenType) -> Style:
        """Get a style for a given Pygments token."""
        raise NotImplementedError  # pragma: no cover

    @abstractmethod
    def get_background_style(self) -> Style:
        """Get the background color."""
        raise NotImplementedError  # pragma: no cover
'''


console = Console()

syntax = Syntax(code, "python", theme="default")


class SyntaxLine:
    def __init__(self, segments):
        self.segments = segments

    def __rich_console__(self, console, options):
        yield from self.segments


lines = [SyntaxLine(segments) for segments in console.render_lines(syntax)]

table = Table("Line No", "Foo", "Line")
for line_no, line in enumerate(lines, 1):
    table.add_row(str(line_no), "f" * line_no, line)

print(table)
```

btw you might want to pin Rich to a more recent version. There have been a lot of updates!

**Reply by emeryberger** [reply]
Just incorporated this into Scalene and it works like a charm - thanks!! (Also, will update the version!)

<img width="1056" alt="Screen Shot 2021-01-27 at 12 12 44 PM" src="https://user-images.githubusercontent.com/1612723/106027873-4e02c800-6099-11eb-8175-82aba77841c1.png">
